### ISSUE TITLE
Should Symbol be an enum or an abstract class?  Should colorless mana be a Symbol?

### ISSUE THREAD

**Issue by catascopic** [c0d3br34k3r/magic#4]
Okay, this one I've thought about entirely too much.  I think there are basically two problems.

The first is a conflict between wanting symbols to be an enum, and wanting them to have a type hierarchy.  Right now I do both by having a symbol enum delegate to an inner symbol class that does have a hierarchy, but the useful things that the hierarchy has to offer are hidden.  For example, I could have a type `Monocolored` which `Primary`, `MonocoloredHybrid`, and `Phyrexian` inherit, but `Hybrid` and `Variable` do not.

If I do want to have a hierarchy, though, I could use an enum-like structure, such as:

```
public static final PrimarySymbol WHITE = new PrimarySymbol...
public static final HybridSymbol WHITE_BLUE = new HybridSymbol...
```

But then to iterate along these values, I'd need to do something I don't like: make an `ImmutableList`, and either re-type out all the values, or use reflection to look for `final static Symbol` fields in the class, and add them to the list.

Also, it would be nice if each subtype of symbol could be declared in its own class, like have class `Primary` with constants `WHITE, BLUE, BLACK, RED, GREEN`, and then have class `Phyrexian` with the same named constants, differentiated by their class name.  This would cut down on the boilerplate prefixes I currently use. But since I need a way to parse symbols from a String, I do actually need a collection somewhere of all the symbols, and this would make it even harder.

And then we have the problem of colorless mana.  I don't really like that colorless is not a symbol, but as long as Symbol is an enum, I think this is the way to go.  I'm not defining all 16 constant colorless symbols in the enum, that seems ridiculous.  If Symbol were an abstract class, it might be okay to have a `ColorlessSymbol` class with a public constructor/factory, to make things less confusing.

But it all comes back to the original problem: Making symbol an enum makes a few things (mostly iteration, which is very important) very easy, but prevents me from doing some other pretty useful stuff.

**Comment by catascopic** [comment]
I have two possible solutions:

Solution 1: [Numbers are a type of mana symbol. ManaCost has an `Optional<NumericSymbol>` and a `Multiset<RepeatableSymbol>`](https://github.com/c0d3br34k3r/magic/tree/feature/colorless_symbol/src/magic)
- Represents symbols as they appear on the card.
- More elegant way of representing monocolored hybrid symbols.
- Since like 90% of cards have a mana cost with one or fewer types of non-numeric symbols, ImmutableMultiset can use its special-case implementations (singleton and empty), which are smaller and faster.
- No special case for printing numeric symbols, just call toString on them.
- The Symbol.parse(String) function works like you'd expect, if you give it a numeric symbol representation, you get a numeric symbol.
- No need to use the "SpecialManaCost" implementation for empty and zero costs.

Solution 2: [Numbers can be an element of a hybrid symbol, but Generic is now a repeatable symbol.  ManaCost simply has a `Multiset<Symbol>`](https://github.com/c0d3br34k3r/magic/tree/feature/colorless_singleton/src/magic)
- Represents the logic of mana payment better.
- Internal storage is easier to work with.
- We don't have to do two different things (other solution makes you check the Numeric and Repeatable symbols) every time we want to do a calculation, such as converted mana cost.
- Formatting to a String is slightly easier, because we can simply sort the symbols into the right order in the Multiset, rather than inserting the Numeric symbol into the right space.
- Reduced rule checking for colorless symbols (originally had to make sure there was only one).

**Comment by murgatroid99** [comment]
Here's another solution:
- `Symbol` is an interface, with methods `convertedManaCost`, `getColors`, `devotionTo(color)`, and probably others.
- Implementors of `Symbol` are `GenericSymbol` (no color), `ColoredSymbol` (one color), `HybridSymbol` (two colors), `MonoColoredHybrid` (one color, 2 cmc), `PhyrexianSymbol` (one color, payable by any), etc.
- `ManaCostElement` would contain a `Symbol` and an `int` (the count).
- `ManaCost` would contain a list of `ManaCostElement`.

This has the following benefits:
- No need to use the "SpecialManaCost" implementation for empty and zero costs. (empty cost=empty list, zero cost=`[ManaCostElement(Generic, 0)]`)
- Basically all of the benefits of Solution 2
